### DISCUSSION TITLE
nullif(x,x) == null?

### DISCUSSION THREAD

**Discussion by RobbertDM** [dbt-labs/dbt-core#6779] [Q&A] [answered]
Hi there! Sometimes `nullif` is used with twice the same value, for example in https://github.com/dbt-labs/dbt-core/blob/main/core/dbt/include/global_project/macros/materializations/snapshots/helpers.sql#L62

Wouldn't that always evaluate to `null`? Why not put `null` instead?

**Comment by dbeatty10** [comment, marked as answer]
Those are some very sharp eagle eyes @RobbertDM ! 🦅

Caveat: all the commentary that follows is my best guess from reading the relevant code -- I didn't actually test any of this out to be certain.

### TL;DR
In the initial snapshot, `dbt_valid_to` will be `null` for all rows. This `nullif` trick forces the cast to the same data type as `dbt_valid_from` rather than the database choosing some other unintended/incompatible data type.

### `null` sounds about right
I think the answer is "yes!" it will always evaluate to `null`. But I suspect that it isn't just _any_ `null`, it's a `null` with the same data type as the `strategy.updated_at` column/expression.

So if `strategy.updated_at` has a datatype of `timestamp`, then I'm thinking it would be equivalent to this:
```
cast(null as timestamp) as dbt_valid_to
```

### More detail
Looking at the [initial commit](https://github.com/dbt-labs/dbt-core/commit/7d490d4886daab2c16c2a1d517f86a27a8f5dac9) for this logic more closely, it looks to me like `nullif({{ strategy.updated_at }}, {{ strategy.updated_at }}) as dbt_valid_to` replaced `{{ timestamp_column.literal('null') }} as dbt_valid_to`.

[Expanding out `literal('null')`](https://github.com/dbt-labs/dbt-core/blob/0ef9931d197c996ccd4e3caf641729176eceda99/core/dbt/adapters/base/column.py#L104-L105
), we'd get something like `null::TIMESTAMP` or `cast(null as TIMESTAMP)` depending on the database adapter.

Why not just do something like `cast(null as timestamp)` directly? Because `strategy.updated_at` can be an expression, dbt might not know the actual data type it has. So the `nullif()` approach is a clever way of handing it to the database and saying "you figure it out!".

**Reply by RobbertDM** [reply]
Oh wow, I didn't even know `null`s had a type :astonished: 
Thanks for explaining!
